orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,568 @@
+        // 1. The size of arg_not_in_subgraph->Shape is smaller than arg_in_subgraph->Shape by 2,
+        //    do not need to add flatten pattern to arg_not_in_subgraph.
+        // 2. The size of arg_not_in_subgraph->Shape is same with arg_in_subgraph->Shape and the first
+        //    two dims value are exactly same, then there are also two possibilities:
+        //     <1>. The arg_not_in_subgraph is propagated from embedding_node (contained in candidate_node_args),
+        //          do not need to process it.
+        //     <2>. The arg_not_in_subgraph is not propagated from embedding_node (not contained in candidate_node_args),
+        //          need to add flatten pattern to arg_not_in_subgraph.
+        //     Here we just add cur node to candidate_input and process it (add flatten pattern to its input) after
+        //     the graph iteration, according to whether it's contained in candidate_node_args.
+        if (arg_in_subgraph->Shape()->dim_size() == arg_not_in_subgraph->Shape()->dim_size()) {
+          candidate_input.insert(cur);
+        }
+      } else {
+        LOG_DEBUG_INFO(logger, "PaddingElimination::Input shapes of node:" + cur->Name() + "are not compatible.");
+        candidate_output.insert(cur);
+        continue;
+      }
+    } else if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "LayerNormalization", {1, 17}, kOnnxDomain)) {
+      if (candidate_node_args.find(cur->MutableInputDefs()[0]) == candidate_node_args.end()){
+        LOG_DEBUG_INFO(logger, "PaddingElimination::Fist input of Normalization: " + cur->Name() +
+                               " is not in candidate_node_args.");
+        candidate_output.insert(cur);
+        continue;
+      }
+      auto axis = static_cast<int64_t>(cur->GetAttributes().at("axis").i());
+      axis = axis < 0 ? axis + cur->InputDefs()[0]->Shape()->dim_size() : axis;
+      if (axis < 2) {
+        LOG_DEBUG_INFO(logger, "PaddingElimination::axis of Normalization: " + cur->Name() + " is " +
+                                   std::to_string(axis) + ", which blocks merging leading two dims.");
+        candidate_output.insert(cur);
+      } else {
+        candidate_node_args.insert(cur->MutableOutputDefs()[0]);
+        PushAllOutputNode(graph, to_visit, cur, visited);
+      }
+    } else if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "Dropout", {12, 13})) {
+      ORT_ENFORCE(candidate_node_args.find(cur->MutableInputDefs()[0]) != candidate_node_args.end());
+      candidate_node_args.insert(cur->MutableOutputDefs()[0]);
+      candidate_node_args.insert(cur->MutableOutputDefs()[1]);
+      PushAllOutputNode(graph, to_visit, cur, visited);
+    } else if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "Cast", {9, 13}) ||
+               graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "BiasGelu", {1}, kMSDomain)) {
+      ORT_ENFORCE(candidate_node_args.find(cur->MutableInputDefs()[0]) != candidate_node_args.end());
```

Comment:
BiasGelu has two inputs?

Reply:
Is it possible the input[1] of BiasGelu is propagated from embedding?